File: DSAlgorithms/patterns/sliding_window/4_longest_substring_k_distinct.ts
const longest_substring_k_distinct = (str: string, k: number) =>
{
    let window_start = 0;
    let max_length = 0;
    let char_frequency = new Map<string, number>();
    for(let window_end = 0; window_end < str.length; window_end++)
    {
        const right_char = str[window_end];
        if(!char_frequency.has(right_char))
            char_frequency.set(right_char, 0);
        else
        {
            const current = char_frequency.get(right_char);
            if(current !== undefined)
                char_frequency.set(right_char, current + 1);
        }
        while(char_frequency.size > k)
        {
            const left_char = str[window_start];
            if(char_frequency.has(left_char) && char_frequency.get(left_char) === 0)
                char_frequency.delete(left_char);
            if(char_frequency.has(left_char))
            {
                const current = char_frequency.get(left_char);
                if(current)
                    char_frequency.set(left_char, current - 1);
            }
            window_start ++; 
        }
        max_length = Math.max(max_length, window_end - window_start + 1);
    }
    return max_length;
}

console.log(`Length of longest substring: ${longest_substring_k_distinct('araaci', 2)}`);
console.log(`Length of longest substring: ${longest_substring_k_distinct('araaci', 1)}`);
console.log(`Length of longest substring: ${longest_substring_k_distinct('cbbebi', 3)}`);

File: DSAlgorithms/patterns/sliding_window/4_longest_substring_k_distinct.cpp
#include <iostream>
#include <bits/stdc++.h>
#include <unordered_map>

class longest_substring_k_distinct {
    public:
        static int find_length(const std::string &str, int k)
        {
            int window_start = 0;
            int max_length = 0;
            std::unordered_map<char, int> char_freq_map;
            for(int window_end = 0; window_end < str.length(); window_end++)
            {
                char right_char = str[window_end];
                char_freq_map[right_char]++;
                while((int)char_freq_map.size() > k)
                {
                    char left_char = str[window_start];
                    char_freq_map[left_char]--;
                    if(char_freq_map[left_char]== 0)
                        char_freq_map.erase(left_char);
                    window_start++;
                }
                max_length = std::max(max_length, window_end - window_start + 1);
            }

            return max_length;
        }
};

int main(int argc, char *argv[])
{
    std::cout<<"Length of longest substring: "<< longest_substring_k_distinct::find_length("araaci", 2);
    std::cout<<'\n';
    std::cout<<"Length of longest substring: "<< longest_substring_k_distinct::find_length("araaci", 1);
    std::cout<<'\n';
    std::cout<<"Length of longest substring: "<< longest_substring_k_distinct::find_length("cbbebi", 3);
    std::cout<<'\n';
}

